### PR TITLE
Reapplies RecItCandidateSet Support and fixes the deployment breaks.

### PR DESCRIPTION
**Pull request by iwsmith**
This re-applies https://github.com/Pocket/recommendation-api/pull/218 and fixes some deployment issues. I flagged the changes between these two commits since you can't really tell from this what was done.

### CODE REVIEW

File: app/graphql/graphql.py [hunk 1]
```diff
@@ -45,7 +45,7 @@ async def resolve_get_slate(self, info, slate_id: str, recommendation_count: int
                                           recommendation_count=recommendation_count)
 
     async def resolve_list_slates(self, info, recommendation_count: int) -> [SlateModel]:
-        return await SlateModel.get_all(recommendation_count=recommendation_count)
+        return await SlateModel.get_all(user_id=info.context.get('user_id'), recommendation_count=recommendation_count)
```
Comment (iwsmith) [inline app/graphql/graphql.py:48]: 🎏

File: app/models/candidate_set.py [hunk 2]
```diff
@@ -88,3 +165,11 @@ async def _query_by_id(cs_id: str) -> Dict[str, Any]:
             response = await table.query(KeyConditionExpression=key_condition)
 
         return response
+
+
+def candidate_set_factory(candidate_set_id: str) -> Union[Type[RecItCandidateSet], Type[DynamoDBCandidateSet]]:
```
Comment (iwsmith) [inline app/models/candidate_set.py:170]: 🎏

File: app/models/recommendation.py [hunk 3]
```diff
@@ -73,14 +73,15 @@ async def get_recommendations(topic_id: str, recommendation_type: Recommendation
...
+    async def get_recommendations_from_experiment(experiment: SlateExperimentModel, user_id: str) -> ['RecommendationModel']:
         """
         Retrieves a list of RecommendationModel objects for on the given slate experiment.
         :param experiment: a SlateExperimentModel instance
+        :param user_id: ID of the user to generate recommendations for
         :return: a list of RecommendationModel instances
         """
         # for each candidate set id, get the candidate set record from the db
-        candidate_sets = await gather(*(CandidateSetModel.get(cs_id) for cs_id in experiment.candidate_sets))
+        candidate_sets = await gather(*(candidate_set_factory(cs_id).get(cs_id, user_id) for cs_id in experiment.candidate_sets))
```
Comment (iwsmith) [inline app/models/recommendation.py:84]: 🎏

File: app/models/candidate_set.py [hunk 4]
```diff
@@ -21,38 +23,113 @@ class CandidateSetModel(BaseModel):
...
             return False
 
         return True
 
-
     @staticmethod
-    @xray_recorder.capture_async('models.candidate_set.get')
-    async def get(cs_id: str) -> 'CandidateSetModel':
+    @xray_recorder.capture_async('models.dynamodb_candidate_set.get')
+    async def get(cs_id: str, user_id: str = None) -> 'DynamoDBCandidateSet':
```
Comment (iwsmith) [inline app/models/candidate_set.py:118]: 🎏

**Comment by iwsmith** [comment]
Here is the output of `git diff` for the original branch and this one:
```diff --git a/app/graphql/graphql.py b/app/graphql/graphql.py
index 8eb3902..0b83837 100644
--- a/app/graphql/graphql.py
+++ b/app/graphql/graphql.py
@@ -45,7 +45,7 @@ class Query(ObjectType):
                                           recommendation_count=recommendation_count)
 
     async def resolve_list_slates(self, info, recommendation_count: int) -> [SlateModel]:
-        return await SlateModel.get_all(recommendation_count=recommendation_count)
+        return await SlateModel.get_all(user_id=info.context.get('user_id'), recommendation_count=recommendation_count)
 
     async def resolve_get_slate_lineup(self, info, slate_lineup_id: str, recommendation_count: int = 10,
                                        slate_count: int = 8) -> SlateLineupModel:
diff --git a/app/main.py b/app/main.py
index 9feb0e4..d82bcd7 100644
--- a/app/main.py
+++ b/app/main.py
@@ -96,7 +96,7 @@ async def load_slate_configs():
             for experiment in slate_config.experiments:
                 for cs in experiment.candidate_sets:
                     logging.info(f"Validating candidate set {cs}")
-                    csm = await candidate_set_factory(cs, None)
+                    csm = candidate_set_factory(cs)
                     if not await csm.verify_candidate_set(cs):
                         # Send event to Sentry, but don't raise it, because missing candidate sets should not
                         # block successfully starting the application.
diff --git a/app/models/candidate_set.py b/app/models/candidate_set.py
index dc16f24..e9d25c4 100644
--- a/app/models/candidate_set.py
+++ b/app/models/candidate_set.py
@@ -6,7 +6,7 @@ import logging
 from aws_xray_sdk.core import xray_recorder
 from boto3.dynamodb.conditions import Key
 from pydantic import BaseModel
-from typing import List, Dict, Any, Union
+from typing import List, Dict, Any, Union, Type
 
 from app.config import dynamodb as dynamodb_config, recit as recit_config
 import app.cache
@@ -115,7 +115,7 @@ class DynamoDBCandidateSet(CandidateSetModel):
 
     @staticmethod
     @xray_recorder.capture_async('models.dynamodb_candidate_set.get')
-    async def get(cs_id: str) -> 'DynamoDBCandidateSet':
+    async def get(cs_id: str, user_id: str = None) -> 'DynamoDBCandidateSet':
         """
         Retrieves a candidate set from the database and instantiates a CandidateSetModel
 
@@ -167,9 +167,9 @@ class DynamoDBCandidateSet(CandidateSetModel):
         return response
 
 
-@xray_recorder.capture_async('models.candidate_set.candidate_set_factory')
-async def candidate_set_factory(candidate_set_id: str, user_id: str) -> Union[DynamoDBCandidateSet, RecItCandidateSet]:
+def candidate_set_factory(candidate_set_id: str) -> Union[Type[RecItCandidateSet], Type[DynamoDBCandidateSet]]:
+    """Given a `candidate_set_id`, return the class that can fetch those candidates."""
     if candidate_set_id.startswith(RECIT_PREFIX):
-        return await RecItCandidateSet.get(candidate_set_id, user_id)
+        return RecItCandidateSet
     else:
-        return await DynamoDBCandidateSet.get(candidate_set_id)
+        return DynamoDBCandidateSet
diff --git a/app/models/recommendation.py b/app/models/recommendation.py
index 7976b0a..6f07a29 100644
--- a/app/models/recommendation.py
+++ b/app/models/recommendation.py
@@ -77,10 +77,11 @@ class RecommendationModel(BaseModel):
         """
         Retrieves a list of RecommendationModel objects for on the given slate experiment.
         :param experiment: a SlateExperimentModel instance
+        :param user_id: ID of the user to generate recommendations for
         :return: a list of RecommendationModel instances
         """
         # for each candidate set id, get the candidate set record from the db
-        candidate_sets = await gather(*(candidate_set_factory(cs_id, user_id) for cs_id in experiment.candidate_sets))
+        candidate_sets = await gather(*(candidate_set_factory(cs_id).get(cs_id, user_id) for cs_id in experiment.candidate_sets))
 
         recommendations = []
         # get the recommendations
diff --git a/app/models/slate.py b/app/models/slate.py
index 11b5743..88470a2 100644
--- a/app/models/slate.py
+++ b/app/models/slate.py
@@ -38,10 +38,11 @@ class SlateModel(BaseModel):
 
     @staticmethod
     @xray_recorder.capture_async('models_slate_get_all')
-    async def get_all(user_id:str, recommendation_count: Optional[int] = 10) -> List['SlateModel']:
+    async def get_all(user_id: str, recommendation_count: Optional[int] = 10) -> List['SlateModel']:
         """
         Retrieves all slates from the database
         :param recommendation_count: int, 0 = no recs, > 0 include this many recs
+        :param user_id: ID of user we are getting slates for.
         :return: a list fo SlateModel objects
         """
         slate_configs = SlateConfigModel.SLATE_CONFIGS_BY_ID.values()
```

**Review by kkyeboah — APPROVED** [review]
👍 and 🤞 for 🎬 2️⃣

**Review by chelseatroy — APPROVED** [review]
We paired on this 🍐

**Pull request merged**